### ISSUE TITLE
MSL: OpGroupNonUniformAllEqual erroneously calls simd_broadcast_first()

### ISSUE THREAD

**Issue by toldjuuso** [KhronosGroup/SPIRV-Cross#1357]
Hi, thanks for fixing the first error. I have now found something else:

Setup: same specs as in #1343

Consider OpGroupNonUniformAllEqual and its specification:

> Evaluates a value for all active invocations in the group. The result is true if Value is equal for all active invocations in the group. Otherwise, the result is false.

This generates **two functions**. The SPIR-V to generate this is:

```
    OpMemoryModel Logical Vulkan
    OpEntryPoint GLCompute %main "main" %invocation_id %_
    OpExecutionMode %main LocalSize 1024 1 1

...


    %main = OpFunction %void None %11
    %16 = OpLabel

        %inv = OpAccessChain %wg %invocation_id %uint_0

        %nil = OpGroupNonUniformAllEqual %bool %uint_3 %float_0
        %newv = OpSelect %float %nil %float_1 %float_0
        %this = OpAccessChain %_ptr_Uniform_float %_ %uint_0 %inv
        OpStore %this %newv


    OpReturn
    OpFunctionEnd
```

Yet, `spirv-cross` with ` --msl --msl-version 020200 ` generates MSL:

```
template<typename T>
inline bool spvSubgroupAllEqual(T value)
{
    return simd_all(value == simd_broadcast_first(value));
}

template<>
inline bool spvSubgroupAllEqual(bool value)
{
    return simd_all(value) || !simd_any(value);
}

kernel void main0(device _7& _4 [[buffer(0)]], uint3 gl_GlobalInvocationID [[thread_position_in_grid]])
{
    _4._m0[gl_GlobalInvocationID.x] = float(spvSubgroupAllEqual(0.0));
}
```

Focus is at **float(spvSubgroupAllEqual(0.0));**. This seems to erroneously call the first declaration of `spvSubgroupAllEqual(T value)`. Indeed, consider fulling the buffer with the following (with f32 types, my subgroup size is 64):

`64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 64 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 `

The result is:

`[1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]`

This, to me, confirms that the wrong function is a) generated and b) called.

Thanks for the project, as always.

**Comment by cdavis5e** [comment]
Are you sure the wrong function is being called? The second overload is for `bool` *arguments*; that is, when the value passed to `OpGroupNonUniformAllEqual` is *itself* a `bool`, which it isn't in this case.

What this operation does is check if all the invocations in the group have the same value for its parameter. There's no native operation for this MSL, so I implemented this by comparing every value to the first invocation's value. If all the comparisons return `true`, then by the transitive property of equality, they should all be equal to each other. Here you passed a constant 0.0 here, so of course it returned `true`.

**Comment by toldjuuso** [comment]
Well, this is embarrassing, you are certainly correct. Somewhere I made a logical mistake and assumed I was using a different opcode: I was *supposed* to use OpGroupNonUniformAll, which takes in a predicate, but then I must have made an error and accidentally switched the command in my source code to OpGroupNonUniformAllEqual, still assuming I was using the logic of OpGroupNonUniformAll. This bias even passed me so badly I did not notice it when filing for the issue -- the initial idea was comparing a register value to a zero-value float.

Also, thanks for the explanation of the code generation -- I was quite surprised when the code generated two functions, but now I see what is the idea behind it.
